packages/dbml-core/src/parse/dbml/src/lib/parser/parser.ts

```diff
@@ -0,0 +1,827 @@
+import { ParsingError, ParsingErrorCode } from '../errors';
+import { SyntaxToken, SyntaxTokenKind, isOpToken } from '../lexer/tokens';
+import Result from '../result';
+import { ParsingContext, ParsingContextStack } from './contextStack';
+import {
+  AttributeNode,
+  BlockExpressionNode,
+  CallExpressionNode,
+  ElementDeclarationNode,
+  ExpressionNode,
+  FieldDeclarationNode,
+  FunctionApplicationNode,
+  FunctionExpressionNode,
+  GroupExpressionNode,
+  InfixExpressionNode,
+  ListExpressionNode,
+  LiteralElementExpressionNode,
+  LiteralNode,
+  NormalFormExpressionNode,
+  PostfixExpressionNode,
+  PrefixExpressionNode,
+  PrimaryExpressionNode,
+  ProgramNode,
+  SyntaxNode,
+  SyntaxNodeKind,
+  TupleExpressionNode,
+  VariableNode,
+} from './nodes';
+
+export default class Parser {
+  private tokens: SyntaxToken[];
+
+  private current: number = 0;
+
+  private errors: ParsingError[] = [];
+
+  private invalid: (SyntaxToken | SyntaxNode)[] = [];
+
+  private contextStack: ParsingContextStack = new ParsingContextStack();
+
+  constructor(tokens: SyntaxToken[]) {
+    this.tokens = tokens;
+  }
+
+  private isAtEnd(): boolean {
+    return (
+      this.current >= this.tokens.length || this.tokens[this.current].kind === SyntaxTokenKind.EOF
+    );
+  }
+
+  private init() {
+    this.current = 0;
+    this.errors = [];
+    this.invalid = [];
+    this.contextStack = new ParsingContextStack();
+  }
+
+  private advance(): SyntaxToken {
+    return this.tokens[this.current++];
+  }
+
+  private peek(lookahead: number = 0): SyntaxToken | undefined {
+    if (lookahead + this.current >= this.tokens.length) {
+      return undefined;
```

Comment:
Instead of returning undefined, could we consider returning the last EOF token instead? By doing this, the return type of our method becomes uniform, and we can eliminate the need for an undefined check every time we invoke it.